extensions/applicationinsights-osplugin-js/src/OsPlugin.ts

```diff
@@ -0,0 +1,257 @@
+} from "@microsoft/applicationinsights-common";
+import {
+    BaseTelemetryPlugin, IAppInsightsCore, IConfigDefaults, IConfiguration, IPlugin, IProcessTelemetryContext,
+    IProcessTelemetryUnloadContext, ITelemetryItem, ITelemetryUnloadState, _eInternalMessageId, _throwInternal, addPageHideEventListener,
+    addPageUnloadEventListener, arrForEach, createProcessTelemetryContext, createUniqueNamespace, eLoggingSeverity, getSetValue,
+    mergeEvtNamespace, onConfigChange, removePageHideEventListener, removePageUnloadEventListener, setValue
+} from "@microsoft/applicationinsights-core-js";
+import { IPromise, doAwaitResponse } from "@nevware21/ts-async";
+import { isString, objDeepFreeze } from "@nevware21/ts-utils";
+import { IOSPluginConfiguration } from "./DataModels";
+
+const defaultMaxTimeout = 200;
+const strExt = "ext";
+interface platformVersionInterface {
+    platform?: string,
+    platformVersion?: string
+}
+interface UserAgentHighEntropyData {
+    platformVersion: platformVersionInterface
+}
+interface ModernNavigator {
+    userAgentData?: {
+      getHighEntropyValues?: (fields: ["platformVersion"]) => IPromise<UserAgentHighEntropyData>;
+    };
+  }
+const defaultOSConfig: IConfigDefaults<IOSPluginConfiguration> = objDeepFreeze({
+    maxTimeout: defaultMaxTimeout,
+    mergeOsNameVersion: undefined
+});
+
+interface IDelayedEvent {
+    ctx: IProcessTelemetryContext;
+    item: ITelemetryItem;
+}
+
+export class OsPlugin extends BaseTelemetryPlugin {
+    public identifier = "OsPlugin";
+    public priority = 195;
+    public version = "#version#";
+
+    constructor() {
+        super();
+        let _core: IAppInsightsCore;
+        let _ocConfig: IOSPluginConfiguration;
+        let _getOSInProgress: boolean;
+        let _getOSTimeout: any;
+        let _maxTimeout: number;
+
+        let _platformVersionResponse: platformVersionInterface;
+        let _retrieveFullVersion: boolean;
+        let _mergeOsNameVersion: boolean;
+
+        let _eventQueue: IDelayedEvent[];
+        let _evtNamespace: string | string[];
+        let _excludePageUnloadEvents: string[];
+
+        let _os: string;
+        let _osVer: number;
+        let _firstAttempt: boolean;
+    
+        dynamicProto(OsPlugin, this, (_self, _base) => {
+
+            _initDefaults();
+
+            _self.initialize = (coreConfig: IConfiguration & IConfig, core: IAppInsightsCore, extensions: IPlugin[]) => {
+                let _self = this;
+                _core = core;
+                super.initialize(coreConfig, core, extensions);
+                let identifier = _self.identifier;
+                _evtNamespace = mergeEvtNamespace(createUniqueNamespace(identifier), core.evtNamespace && core.evtNamespace());
+                if (utlCanUseSessionStorage){
+                    _platformVersionResponse = JSON.parse(utlGetSessionStorage(core.logger, "ai_osplugin"));
```

Comment:
We should wrap this into a try / catch as "someone" could put garbage into the field that causes an exception,